Source/Core/Core/HW/DSPHLE/UCodes/ASnd.cpp

```diff
@@ -0,0 +1,549 @@
+
+// November 14, 2008 version (padded to 0x05a0 bytes) - initial release
+// https://github.com/devkitPro/libogc/compare/c76d8b851fafc11b0a5debc0b40842929d5a5825~...353a44f038e75e5982eb550173ec8127ab35e3e3
+constexpr u32 HASH_2008 = 0x8d69a19b;
+// February 5, 2009 version (0x05ac bytes) - added MAIL_TERMINATE
+// https://github.com/devkitPro/libogc/compare/1925217ffb4c97cbee5cf21fa3c0231029b340e2~...3b1f018dbe372859a43bff8560e2525f6efa4433
+constexpr u32 HASH_2009 = 0xc2fd441c;
+// June 11, 2011 version (padded to 0x0620 bytes) - added new sample formats, which shifted flags
+// Note that the source include in the repo does not match the compiled binary exactly; the compiled
+// version differs by using asl instead of lsl, $acc1 instead of $acc0, and $ac0.l instead of $ac0.m
+// in various locations, as well as having the "jmp out_samp" line uncommented in stereo_16bits_le.
+// None of these result in a behavior difference, from the source, though.
+// Note that gcdsptool was also updated, which results in some differences in the source that don't
+// actually correspond to different instructions (e.g. s40 was renamed to s16)
+// https://github.com/devkitPro/libogc/commit/b1b8ecab3af3745c8df0b401abd512bdf5fcc011
+constexpr u32 HASH_2011 = 0xa81582e2;
+// June 12, 2020 version (0x0606 bytes) - libogc switched to compiling the ucode at build time
+// instead of including a pre-compiled version in the repo, so this now corresponds to the code
+// provided in the repo. There appear to be no behavior differences from the 2011 version.
+// https://github.com/devkitPro/libogc/compare/bfb705fe1607a3031d18b65d603975b68a1cffd4~...d20f9bdcfb43260c6c759f4fb98d724931443f93
+constexpr u32 HASH_2020 = 0xdbbeeb61;
+
+constexpr u32 SAMPLE_RATE = 48000;
+
+ASndUCode::ASndUCode(DSPHLE* dsphle, u32 crc) : UCodeInterface(dsphle, crc)
+{
+}
+
+ASndUCode::~ASndUCode()
+{
+}
+
+void ASndUCode::Initialize()
+{
+  m_mail_handler.PushMail(DSP_INIT);
+}
+
+void ASndUCode::Update()
+{
+  // This is dubious in general, since we set the interrupt parameter on m_mail_handler.PushMail
+  if (!m_mail_handler.IsEmpty())
+  {
+    DSP::GenerateDSPInterruptFromDSPEmu(DSP::INT_DSP);
+  }
+}
+
+void ASndUCode::HandleMail(u32 mail)
+{
+  if (m_upload_setup_in_progress)
+  {
+    PrepareBootUCode(mail);
+  }
+  else if (m_next_command_is_voice_addr)
+  {
+    m_voice_addr = mail;
+    INFO_LOG_FMT(DSPHLE, "ASndUCode - Voice data is at {:08x}", mail);
+    m_next_command_is_voice_addr = false;
+    // No mail is sent in response
+  }
+  else if (mail >> 16 == 0xCDD1)
+  {
+    switch (mail)
+    {
+    case MAIL_NEW_UCODE:
+      m_upload_setup_in_progress = true;
+      break;
+    case MAIL_RESET:
+      m_dsphle->SetUCode(UCODE_ROM);
+      break;
+    default:
+      WARN_LOG_FMT(DSPHLE, "ASndUCode - unknown 0xcdd1 command: {:08x}", mail);
+      break;
+    }
+    // No mail is sent in response to any of these
+  }
+  else
+  {
+    switch (mail & 0xffff)
```

Comment:
Since we use #10753 , we should be able to follow the same logic there.  No need to mask `mail` right?

Reply:
No, this is an accuracy thing.  I have a comment about this:

https://github.com/dolphin-emu/dolphin/blob/d3f082722c67b6c0ebf25fc4b3a97e243466c697/Source/Core/Core/HW/DSPHLE/UCodes/ASnd.cpp#L20

but I should probably move it down to this switch statement instead of next to the constants.